### ISSUE TITLE
[Bug] FSDP2 mixed-precision upcast to fp32 is a silent no-op since v1.13.0

### ISSUE THREAD

**Issue by iavinas** [huggingface/accelerate#3979]
## Bug

In `fsdp2_prepare_model()`, the parameter upcast to fp32 master weights does nothing:

```python
# src/accelerate/utils/fsdp_utils.py (current main)
for name, param in model.named_parameters():
    if param.requires_grad and param.dtype != torch.float32:
        upcasted_params.append(name)
        param = param.to(torch.float32)  # ← BUG: rebinds local variable, model unchanged
```

`param = param.to(...)` creates a new tensor and assigns it to the local variable `param`. The model's actual parameter is untouched. The warning fires ("FSDP upcast of low precision parameters to fp32...") but nothing was upcasted.

**Result:** FSDP2 + mixed precision trains with bf16 master weights instead of fp32. Optimizer updates below bf16 resolution (~0.008) are silently lost. Training is **28x slower to converge** compared to correct fp32 masters (proven with pure PyTorch — see below).

## Introduced by

**PR #3905** (`758f3b18`, merged into v1.13.0) changed:
```diff
-                param.data = param.data.to(torch.float32)
+                param = param.to(torch.float32)
```

The intent was to fix a DTensor crash at `optimizer.step()` — `param.data = ...` doesn't work on DTensor params after `fully_shard`. But the "fix" turned the upcast into a no-op, trading a crash for silent training degradation.

## Proof (pure PyTorch, no Accelerate)

Four experiments with identical model, data, and hyperparameters — only the upcast strategy differs:

```
torchrun --nproc_per_node=2 repro_fsdp2_pure_pytorch.py
```

```
=================================================================
  COMPARISON (lr=1e-4, 50 steps, y=2x+1 MSE)
=================================================================
  ✗ A: bf16+policy (no upcast)        4.9971 → 4.9951  Δ=0.0020
  ✗ B: bf16 raw (no policy)           4.9982 → 4.9960  Δ=0.0022
  ✓ C: fp32+policy (gold standard)    4.9971 → 4.9403  Δ=0.0568
  ✓ D: bf16→fp32 BEFORE fully_shard   4.9971 → 4.9395  Δ=0.0576
```

**D ≈ C >> A.** Upcasting before `fully_shard` matches the gold standard. Without upcast, 28x less learning.

Key observations from forward hooks:
- **All four experiments** use `weight=torch.bfloat16` during forward (MixedPrecisionPolicy casts for compute)
- The difference is only in master weight dtype: A/B store bf16, C/D store fp32
- fp32 masters → fp32 optimizer states → small updates accumulate correctly

## Why the benchmarks didn't catch it

`benchmarks/fsdp2/utils.py` loads models in fp32:
```python
model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.float32)
```
The upcast check `param.dtype != torch.float32` was always False — broken code never ran.

## Why a simple revert doesn't work

Reverting to `param.data = param.data.to(fp32)` at the original location (after `fully_shard`) causes two problems:

1. **DTensor limitation:** `param.data = ...` on FSDP2-wrapped DTensor params doesn't properly update the DTensor's cached dtype (verified experimentally — `dt.dtype` stays bf16 even after `dt._local_tensor.data` is mutated to fp32)

2. **FSDP2 dtype snapshot:** `fully_shard()` snapshots param dtype at call time. Upcasting after means FSDP2 recorded bf16 as master dtype → post-backward hooks cast gradients to bf16 → optimizer gets mixed dtypes

## Correct fix

Upcast **before** `fully_shard()` so FSDP2 snapshots fp32:

| Path | Strategy |
|---|---|
| Non-ram-efficient | `param.data = param.data.to(fp32)` before `fully_shard` (params are regular tensors, not DTensors yet) |
| Ram-efficient | Upcast state dict to fp32 + `model.to(device="meta", dtype=fp32)` so broadcast buffers and `fully_shard` both see fp32 |

### Needs guidance: TP + FSDP2 (DTensor params)

The original PR #3905 crash was with TP + FSDP2 where params are DTensors. I confirmed that `param.data = param.data.to(fp32)` on a DTensor doesn't update its cached dtype — so a simple revert of PR #3905 would still break TP + FSDP2.

I don't have a concrete fix for the DTensor case. I'd appreciate maintainer guidance on the correct approach — specifically, what is the expected TP + FSDP2 + mixed precision flow, and where should the upcast happen relative to TP application?

## Reproduction

```bash
pip install accelerate>=1.13.0

# Accelerate repro (2+ GPUs)
accelerate launch --num_processes 2 --mixed_precision bf16 repro_fsdp2_upcast.py

# Pure PyTorch proof (2+ GPUs)
torchrun --nproc_per_node=2 repro_fsdp2_pure_pytorch.py
```

<details>
<summary>repro_fsdp2_upcast.py (Accelerate)</summary>

```python
"""
Minimal FSDP2 upcast diagnostic. Reports master weight dtype and loss delta.

Usage (requires 2+ GPUs):
    accelerate launch --num_processes 2 --mixed_precision bf16 repro_fsdp2_upcast.py
"""

import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset

from accelerate import Accelerator
from accelerate.state import AcceleratorState
from accelerate.utils import FullyShardedDataParallelPlugin, set_seed


class SimpleMLP(nn.Module):
    def __init__(self, input_dim=16, hidden_dim=128, output_dim=16):
        super().__init__()
        self.layer1 = nn.Linear(input_dim, hidden_dim)
        self.relu1 = nn.ReLU()
        self.layer2 = nn.Linear(hidden_dim, hidden_dim)
        self.relu2 = nn.ReLU()
        self.layer3 = nn.Linear(hidden_dim, output_dim)

    def forward(self, x):
        return self.layer3(self.relu2(self.layer2(self.relu1(self.layer1(x)))))


def make_dataset(n_samples=512, input_dim=16):
    torch.manual_seed(123)
    x = torch.randn(n_samples, input_dim)
    y = 2.0 * x + 1.0
    return TensorDataset(x, y)


def run(ram_efficient):
    set_seed(42)
    fsdp_plugin = FullyShardedDataParallelPlugin(fsdp_version=2, cpu_ram_efficient_loading=ram_efficient)
    accelerator = Accelerator(fsdp_plugin=fsdp_plugin)
    is_main = accelerator.is_main_process

    model = SimpleMLP().to(torch.bfloat16)
    optimizer = torch.optim.AdamW(model.parameters(), lr=1e-4)
    criterion = nn.MSELoss()
    dataloader = DataLoader(make_dataset(), batch_size=32, shuffle=False)

    if is_main:
        print(f"\n  ram_efficient={ram_efficient}  mixed_precision={accelerator.mixed_precision}")
        print(f"  BEFORE prepare: {{{next(iter({p.dtype for p in model.parameters()}))}}}")

    model, optimizer, dataloader = accelerator.prepare(model, optimizer, dataloader)

    if is_main:
        print(f"  AFTER  prepare: {{{next(iter({p.dtype for p in model.parameters()}))}}}")

    model.train()
    losses = []
    for step, (bx, by) in enumerate(dataloader):
        if step >= 50:
            break
        loss = criterion(model(bx).float(), by.float())
        losses.append(loss.item())
        accelerator.backward(loss)
        optimizer.step()
        optimizer.zero_grad()

    if is_main:
        for g in optimizer.param_groups:
            for p in g["params"]:
                s = optimizer.state.get(p, {})
                if "exp_avg" in s:
                    print(f"  Optimizer: param={p.dtype}  exp_avg={s['exp_avg'].dtype}")
                    break
            break
        delta = losses[0] - losses[-1]
        print(f"  Loss: {losses[0]:.4f} → {losses[-1]:.4f}  Δ={delta:.4f}")

    AcceleratorState._reset_state(True)
    return losses


if __name__ == "__main__":
    for ram_eff in [False, True]:
        run(ram_eff)
```

</details>

<details>
<summary>repro_fsdp2_pure_pytorch.py (proof without Accelerate)</summary>

```python
"""
Pure PyTorch FSDP2: proves fp32 master weights require explicit upcast.
A = bf16+policy (no upcast), B = bf16 raw (no policy), C = fp32+policy (gold), D = bf16→fp32 upcast before fully_shard.

Usage: torchrun --nproc_per_node=2 repro_fsdp2_pure_pytorch.py
"""

import os

import torch
import torch.distributed as dist
import torch.nn as nn
from torch.distributed.fsdp import MixedPrecisionPolicy, fully_shard
from torch.utils.data import DataLoader, TensorDataset


class SimpleMLP(nn.Module):
    def __init__(self, input_dim=16, hidden_dim=128, output_dim=16):
        super().__init__()
        self.layer1 = nn.Linear(input_dim, hidden_dim)
        self.relu1 = nn.ReLU()
        self.layer2 = nn.Linear(hidden_dim, hidden_dim)
        self.relu2 = nn.ReLU()
        self.layer3 = nn.Linear(hidden_dim, output_dim)

    def forward(self, x):
        return self.layer3(self.relu2(self.layer2(self.relu1(self.layer1(x)))))


def make_dataset(n_samples=512, input_dim=16):
    torch.manual_seed(123)
    x = torch.randn(n_samples, input_dim)
    y = 2.0 * x + 1.0
    return TensorDataset(x, y)


# Forward hook — capture weight dtype during first forward
fwd_dtypes = {}


def register_fwd_hooks(model):
    fwd_dtypes.clear()
    for name, mod in model.named_modules():
        if isinstance(mod, nn.Linear):
            def hook(module, input, output, n=name):
                if n not in fwd_dtypes:
                    fwd_dtypes[n] = module.weight.dtype
            mod.register_forward_hook(hook)


def run(rank, label, model_dtype, upcast_before_shard, use_mp_policy=True):
    torch.manual_seed(42)
    torch.cuda.manual_seed_all(42)
    model = SimpleMLP().to(model_dtype).to(f"cuda:{rank}")
    criterion = nn.MSELoss()
    mp_policy = MixedPrecisionPolicy(
        param_dtype=torch.bfloat16, reduce_dtype=torch.bfloat16, output_dtype=torch.bfloat16,
    ) if use_mp_policy else None

    if rank == 0:
        print(f"  Before: dtypes={{{next(iter({p.dtype for p in model.parameters()}))}}}")

    if upcast_before_shard:
        for name, param in model.named_parameters():
            if param.requires_grad and param.dtype != torch.float32:
                param.data = param.data.to(torch.float32)
        if rank == 0:
            print(f"  After upcast: dtypes={{{next(iter({p.dtype for p in model.parameters()}))}}}")

    fsdp_kwargs = {}
    if mp_policy is not None:
        fsdp_kwargs["mp_policy"] = mp_policy
    fully_shard(model, **fsdp_kwargs)

    if rank == 0:
        print(f"  After shard: dtypes={{{next(iter({p.dtype for p in model.parameters()}))}}}")

    optimizer = torch.optim.AdamW(model.parameters(), lr=1e-4)
    register_fwd_hooks(model)

    dataloader = DataLoader(make_dataset(), batch_size=32, shuffle=False)

    model.train()
    losses = []
    for step, (x, y) in enumerate(dataloader):
        if step >= 50:
            break
        x = x.to(f"cuda:{rank}")
        y = y.to(f"cuda:{rank}")
        if not use_mp_policy:
            x, y = x.to(model_dtype), y.to(model_dtype)
        pred = model(x)
        loss = criterion(pred.float(), y.float())
        losses.append(loss.item())
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()

    if rank == 0:
        fwd_dt = next(iter(fwd_dtypes.values())) if fwd_dtypes else "N/A"
        opt_dt = "N/A"
        for g in optimizer.param_groups:
            for p in g["params"]:
                s = optimizer.state.get(p, {})
                if "exp_avg" in s:
                    opt_dt = s["exp_avg"].dtype
                    break
            break

        delta = losses[0] - losses[-1]
        print(f"  Forward weight: {fwd_dt}  Optimizer: {opt_dt}")
        print(f"  Loss: {losses[0]:.6f} → {losses[-1]:.6f}  Δ={delta:.6f}")

    return losses


def main():
    dist.init_process_group("nccl")
    rank = int(os.environ["LOCAL_RANK"])
    torch.cuda.set_device(rank)

    if rank == 0:
        print("=" * 70)
        print("  Pure PyTorch FSDP2 — y=2x+1, MSE, lr=1e-4, 50 steps")
        print("=" * 70)

    #                (label,                                     dtype,          upcast, policy)
    experiments = [
        ("A: bf16+policy (no upcast = current bug)", torch.bfloat16, False, True),
        ("B: bf16 raw (no policy)",                  torch.bfloat16, False, False),
        ("C: fp32+policy (gold standard)",           torch.float32,  False, True),
        ("D: bf16→fp32 BEFORE fully_shard (fix)",    torch.bfloat16, True,  True),
    ]

    results = {}
    for label, model_dtype, upcast, policy in experiments:
        if rank == 0:
            print(f"\n{'─' * 70}")
            print(f"  {label}")
            print(f"{'─' * 70}")
        losses = run(rank, label, model_dtype, upcast, use_mp_policy=policy)
        results[label] = losses
        dist.barrier()

    if rank == 0:
        print(f"\n{'=' * 70}")
        print("  SUMMARY")
        print(f"{'=' * 70}")
        for label, losses in results.items():
            delta = losses[0] - losses[-1]
            marker = "✓" if delta > 0.01 else "~"
            print(f"  {marker} {label:45s} {losses[0]:.4f} → {losses[-1]:.4f}  Δ={delta:.4f}")

    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

</details>

## Environment

- accelerate: v1.13.0 (current `main` still affected)
- Reproduced on: 2×T4, 4×RTX A4000

**Comment by yuxinyuan** [comment]
Is this fixed ?
